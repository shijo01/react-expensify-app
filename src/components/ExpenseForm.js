import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'
import 'react-dates/initialize';

const now = moment();

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: undefined,
            buttonText: props.expense ? 'Save Expense' : 'Add Expense'
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));

    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt)
            this.setState(() => ({createdAt}));
    };

    onCalenderFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please provide description and amount'}));
        } else {
            this.setState(() => ({error: undefined}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }

    };

    render() {
        return (
            <form
                className='form'
                onSubmit={this.onSubmit}
            >
                {this.state.error && <p className='form__error'>{this.state.error}</p>}
                <input
                    type='text'
                    placeholder='Description'
                    className='text-input'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className='text-input'
                    type='number'
                    placeholder='Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onCalenderFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => {
                        return false;
                    }}
                />
                <textarea
                    className='text-area'
                    onChange={this.onNoteChange}
                    placeholder='Add a note for your expense(optional)'
                    value={this.state.note}
                >

                    </textarea>
                <div>
                    <button className='login-button'>{this.state.buttonText}</button>
                </div>

            </form>
        );
    }

}