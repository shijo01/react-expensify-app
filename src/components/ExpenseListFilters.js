import React from "react"
import {connect} from 'react-redux'
import {setEndDate, setStartDate, setTextFilter, sortByAmount, sortByDate} from "../actions/filters";
import {DateRangePicker, SingleDatePicker} from "react-dates";
import 'react-dates/initialize'

class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };

    onDatesChange = ({startDate, endDate} = {}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({calenderFocused}));
    };

    render() {
        return (
            <div className="content-container">
                <div className="filter-group">
                    <div className='filter-group__item'>
                        <input
                            className="text-input"
                            placeholder='Search expenses'
                            type="text"
                            value={this.props.filters.text}
                            onChange={
                                (e) => {
                                    this.props.dispatch(setTextFilter(e.target.value));
                                }
                            }
                        />
                    </div>
                    <div className='filter-group__item'>
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={
                                (e) => {
                                    if (e.target.value === 'date') {
                                        this.props.dispatch(sortByDate());
                                    } else if (e.target.value === 'amount') {
                                        this.props.dispatch(sortByAmount())
                                    }
                                }
                            }
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>

                    </div>
                    <div className='filter-group__item'>
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            startDateId={"0"}
                            endDateId={"1"}
                            numberOfMonths={2}
                            isOutsideRange={() => false}
                            showClearDates={true}

                        />
                    </div>
                </div>
            </div>
        );
    }

}


const mapStateToProps = (state) => {
    return {
        filters: state.filter
    }
};

export default connect(mapStateToProps)(ExpenseListFilters);
