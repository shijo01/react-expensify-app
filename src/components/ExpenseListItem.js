import React from 'react';
import {NavLink} from "react-router-dom";
import {connect} from 'react-redux';
import moment from 'moment';
import numeral from 'numeral';

numeral.register('locale', 'in', {
    delimiters: {
        thousands: ',',
        decimal: '.'
    },
    abbreviations: {
        thousand: 'K',
        million: 'M',
        billion: 'B',
        trillion: 'T'
    },
    ordinal : function (number) {
        return number === 1 ? 'er' : 'ème';
    },
    currency: {
        symbol: '₹'
    }
});

numeral.locale('in');

export const ExpenseListItem = ({dispatch, id, amount, description, createdAt} = {}) => (
    <div>
        <NavLink to={'/edit/' + id}>
            <h3>{description}</h3>
        </NavLink>
        <p>
            {numeral(amount/100).format('$0,0.00')}
            -
            {moment(createdAt).format('MMM Do, YYYY')}
        </p>
    </div>
);


export default ExpenseListItem;
