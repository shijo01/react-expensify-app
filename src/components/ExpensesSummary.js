import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({expenseCount, expensesTotal} = {}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : "expenses";
    const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpenseTotal}</h1>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filter);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }

};

export default connect(mapStateToProps)(ExpenseSummary);

