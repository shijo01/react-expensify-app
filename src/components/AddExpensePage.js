import React from "react";
import ExpenseForm from "./ExpenseForm";
import {connect} from 'react-redux'
import {startAddExpense} from "../actions/expenses";

const AddExpensePage = (props) => {
    return (
        <div>
            <div className='page-header'>
                <div className='content-container'>
                    <h1 className='page-header__title'>Add Expense</h1>
                </div>
            </div>
            <div className='content-container'>
                <ExpenseForm onSubmit={(expense) => {
                    props.dispatch(startAddExpense(expense));
                    props.history.push('/');
                }}/>
            </div>
        </div>
    );
};


export default connect()(AddExpensePage);

