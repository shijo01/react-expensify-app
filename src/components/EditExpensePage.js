import React from "react";
import {connect} from 'react-redux'
import ExpenseForm from "./ExpenseForm";
import {editExpense, removeExpense, startEditExpense, startRemoveExpense} from "../actions/expenses";

const EditExpensePage = (props) => {
    return (
        <div>
            <div className='page-header'>
                <div className='content-container'>
                    <h1 className='page-header__title'>Modify Expense</h1>
                </div>
            </div>
            <div className='content-container'>
                <ExpenseForm
                    expense={props.expense}
                    onSubmit={(expense) => {
                        props.dispatch(startEditExpense({
                            id: props.expense.id,
                            updates: expense
                        }));
                        props.history.push('/');
                    }}
                />

                <button className='login-button login-button--remove'
                        onClick={
                            () => {
                                props.dispatch(startRemoveExpense(props.expense.id));
                                props.history.push('/');
                            }
                        }>
                    Delete Expense
                </button>
            </div>


        </div>
    )
};
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }

}
export default connect(mapStateToProps)(EditExpensePage);