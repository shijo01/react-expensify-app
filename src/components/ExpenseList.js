import {connect} from 'react-redux'
import React from 'react'
import ExpenseListItem from "./ExpenseListItem";
import expensesSelector from '../selectors/expenses'

export const ExpenseList = (props) => {
    return (
        <div className='content-container'>
            <div className='list-header'>
                <div className='show-for-mobile'>Expenses</div>
                <div className="show-for-larger-screen">Expense</div>
                <div className="show-for-larger-screen">Amount</div>
            </div>
            <div className='list-body'>
                {
                    props.expenses.length === 0 ? (
                        <div className='list-item list-item--message'>
                            <p>No expenses</p>
                        </div>

                    ) : (
                        props.expenses.map((expense) => {
                            return <ExpenseListItem
                                key={expense.id}
                                {...expense}
                            />
                        })
                    )
                }
            </div>

        </div>
    )
};

const mapStateToProps = (state) => ({
        expenses: expensesSelector(state.expenses, state.filter)
    }
);

export default connect(mapStateToProps)(ExpenseList);
