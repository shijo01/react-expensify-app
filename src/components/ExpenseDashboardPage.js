import React from "react";
import ExpenseList from './ExpenseList'
import ExpenseListFilters from "./ExpenseListFilters"

const ExpenseDashboardPage = () => (
    <div>
        Expense Dashboard page
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
);

export default ExpenseDashboardPage;