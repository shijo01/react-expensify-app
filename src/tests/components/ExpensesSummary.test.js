import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSummary} from '../../components/ExpensesSummary';

test('should correctly render expenses summary with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={19000}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render expenses summary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={13} expensesTotal={45967000}/>);
    expect(wrapper).toMatchSnapshot();
});