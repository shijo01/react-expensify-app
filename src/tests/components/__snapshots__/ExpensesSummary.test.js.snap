// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $190.00
  </h1>
</div>
`;

exports[`should correctly render expenses summary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    13
     
    expenses
     totalling 
    $459,670.00
  </h1>
</div>
`;
